app/src/main/java/com/duckduckgo/app/browser/tabs/DefaultTabManager.kt

```diff
@@ -121,37 +155,59 @@ class RealTabManager @Inject constructor(
         }
     }
 
+    override fun clearTabsInMemory() {
+        tabPagerAdapter.clearFragments()
+    }
+
     override fun onCleanup() {
         openMessageInNewTabJob?.cancel()
     }
 
-    private fun selectTab(tab: TabEntity?) {
-        if (swipingTabsFeature.isEnabled) {
-            return
-        }
-
-        Timber.v("Select tab: $tab")
+    private fun requestNewTab(): TabEntity = runBlocking(dispatchers.io()) {
+        val tabId = browserActivity.viewModel.onNewTabRequested()
+        return@runBlocking tabRepository.flowTabs.transformWhile { result ->
+            result.firstOrNull { it.tabId == tabId }?.let { entity ->
+                emit(entity)
+                return@transformWhile true
+            }
+            return@transformWhile false
+        }.first()
+    }
 
-        if (tab == null) return
+    private fun getSelectedTabId(): String? = runBlocking {
+        tabRepository.getSelectedTab()?.tabId
+    }
 
-        if (tab.tabId == currentTab?.tabId) return
+    private fun getTabById(tabId: String): TabEntity? = runBlocking {
+        tabRepository.getTab(tabId)
+    }
 
-        lastActiveTabs.add(tab.tabId)
+    private fun selectTab(tabId: String) = browserActivity.lifecycleScope.launch {
+        if (tabId != currentTab?.tabId) {
+            lastActiveTabs.add(tabId)
 
-        browserActivity.viewModel.onTabActivated(tab.tabId)
+            browserActivity.viewModel.onTabActivated(tabId)
```

Comment:
this smells. The `TabManager` is injected in the `BrowserActivity`, but here you are calling the `ViewModel` from the `BrowserActivity` directly?

Reply:
I’ve refactored this.